flink-connector-debezium/src/main/java/com/alibaba/ververica/cdc/debezium/internal/FlinkDatabaseSchemaHistory.java

```diff
@@ -0,0 +1,205 @@
+ * history DDLs, it's useful to prevent OOM when meet massive history DDLs.
+ */
+public class FlinkDatabaseSchemaHistory implements DatabaseHistory {
+
+    public static final String DATABASE_HISTORY_INSTANCE_NAME = "database.history.instance.name";
+
+    private final JsonTableChangeSerializer tableChangesSerializer =
+            new JsonTableChangeSerializer();
+
+    private ConcurrentMap<TableId, HistoryRecord> tables;
+    private String instanceName;
+    private DatabaseHistoryListener listener;
+    private boolean storeOnlyMonitoredTablesDdl;
+    private boolean skipUnparseableDDL;
+    private boolean useCatalogBeforeSchema;
+
+    @Override
+    public void configure(
+            Configuration config,
+            HistoryRecordComparator comparator,
+            DatabaseHistoryListener listener,
+            boolean useCatalogBeforeSchema) {
+        this.instanceName = config.getString(DATABASE_HISTORY_INSTANCE_NAME);
+        this.listener = listener;
+        this.storeOnlyMonitoredTablesDdl = config.getBoolean(STORE_ONLY_MONITORED_TABLES_DDL);
+        this.skipUnparseableDDL = config.getBoolean(SKIP_UNPARSEABLE_DDL_STATEMENTS);
+        this.useCatalogBeforeSchema = useCatalogBeforeSchema;
+
+        // recover
+        this.tables = new ConcurrentHashMap<>();
+        for (HistoryRecord record : StateUtils.retrieveHistory(instanceName)) {
+            // validate here
+            TableChange tableChange =
+                    JsonTableChangeSerializer.fromDocument(
+                            record.document(), useCatalogBeforeSchema);
+            tables.put(tableChange.getId(), record);
+        }
+        // register
+        StateUtils.registerHistory(instanceName, tables.values());
+    }
+
+    @Override
+    public void start() {
+        listener.started();
+    }
+
+    @Override
+    public void record(
+            Map<String, ?> source, Map<String, ?> position, String databaseName, String ddl)
+            throws DatabaseHistoryException {
+        throw new UnsupportedOperationException(
+                "The FlinkDatabaseSchemaHistory needs debezium provides the schema.");
+    }
+
+    @Override
+    public void record(
+            Map<String, ?> source,
+            Map<String, ?> position,
+            String databaseName,
+            String schemaName,
+            String ddl,
+            TableChanges changes)
+            throws DatabaseHistoryException {
+        final HistoryRecord record =
+                new HistoryRecord(source, position, databaseName, schemaName, ddl, changes);
+        for (TableChanges.TableChange change : changes) {
+            switch (change.getType()) {
+                case CREATE:
+                case ALTER:
+                    tables.put(
+                            change.getId(),
+                            new HistoryRecord(tableChangesSerializer.toDocument(change)));
```

Comment:
I would like to use `HistoryRecord(Map<String, ?> source, Map<String, ?> position, String databaseName, String schemaName, String ddl, TableChanges changes)` to construct a complete `HistoryRecord`. Your way may have some problems because it's not a valid `HistoryRecord`, because TableChanges is not in "tableChanges" field.